aws-api/src/main/java/com/amplifyframework/api/aws/AWSApiPlugin.java

```diff
@@ -361,65 +363,61 @@ private boolean isReadRestrictingOwner(AuthRule authRule) {
-            // Could not read identity value from the token...
-            // Exception will be thrown so do nothing for now
-        }
-
-        if (identityValue == null || identityValue.isEmpty()) {
             throw new ApiException(
-                    "Attempted to subscribe to a model with owner based authorization without " + identityClaim + " " +
-                    "which was specified (or defaulted to) as the identity claim.",
+                    "Attempted to subscribe to a model with owner-based authorization without " + identityClaim + " " +
+                            "which was specified (or defaulted to) as the identity claim.",
                     "If you did not specify a custom identityClaim in your schema, make sure you are logged in. If " +
                             "you did, check that the value you specified in your schema is present in the access key."
             );
         }
-
-        return identityValue;
     }
 
-    private ArrayList<String> getUserGroups(AuthorizationType authType) throws ApiException {
-        // Custom groups claim isn't supported yet.
-        if (!AuthorizationType.AMAZON_COGNITO_USER_POOLS.equals(authType)) {
-            throw new ApiException("Custom groups claim is not supported yet.",
-                    "Please use Amazon Cognito User Pools to authorize your API.");
-        }
-
+    private ArrayList<String> getUserGroups(String groupClaim, AuthorizationType authType) throws ApiException {
         ArrayList<String> groups = new ArrayList<>();
-        final String GROUPS_KEY = "cognito:groups";
-
         try {
-            JSONObject accessToken = CognitoJWTParser.getPayload(getAuthToken(authType));
-
-            if (accessToken.has(GROUPS_KEY)) {
-                JSONArray jsonGroups = accessToken.getJSONArray(GROUPS_KEY);
-
-                for (int i = 0; i < jsonGroups.length(); i++) {
-                    groups.add(jsonGroups.getString(i));
+            JSONObject accessToken = CognitoJWTParser
+                    .getPayload(getAuthToken(authType));
+            if (accessToken.has(groupClaim)) {
+                JSONArray jsonGroups = accessToken.getJSONArray(groupClaim);
+                for (int index = 0; index < jsonGroups.length(); index++) {
+                    groups.add(jsonGroups.getString(index));
                 }
             }
         } catch (JSONException | CognitoParameterInvalidException error) {
             throw new ApiException(
-                    "Failed to parse groups from auth rule.",
-                    error,
-                    "This should never happen - see attached exception for more details and report to us on GitHub."
+                    "Failed to parse group claim from the token.",
+                    AmplifyException.REPORT_BUG_TO_AWS_SUGGESTION
             );
         }
 
         return groups;
     }
 
+    private boolean userNotInReadRestrictingGroups(
```

Comment:
How can we carve more of this logic out of the `AWSApiPlugin`?

Reply:
this will have to be merge-conflicted in my [other PR](#920), ugh. This helper method will definitely not live in the plugin code for long.

Reply:
Ah right, I forgot that wasn't in, yet. I was like "damn, I thought we moved some of this out already..."

Sounds good. I remember you saying you'd rebase that one onto this. That plan works.